satellite/gc/service.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	monkit "gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/bloomfilter"
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/satellite/metainfo"
+	"storj.io/storj/uplink/piecestore"
+)
+
+var (
+	// Error defines the gc service errors class
+	Error = errs.Class("gc service error")
+	mon   = monkit.Package()
+)
+
+// Config contains configurable values for garbage collection
+type Config struct {
+	Interval time.Duration `help:"the time between each send of garbage collection filters to storage nodes" releaseDefault:"168h" devDefault:"10m"`
+	Enabled  bool          `help:"set if garbage collection is enabled or not" releaseDefault:"false" devDefault:"true"`
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int     `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+	ConcurrentSends   int     `help:"the number of nodes to concurrently send bloom filters to" releaseDefault:"1" devDefault:"1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log    *zap.Logger
+	config Config
+	Loop   sync2.Cycle
+
+	transport    transport.Client
+	overlay      overlay.DB
+	metainfoLoop *metainfo.Loop
+}
+
+// RetainInfo contains info needed for a storage node to retain important data and delete garbage data
+type RetainInfo struct {
+	Filter       *bloomfilter.Filter
+	CreationDate time.Time
+	Count        int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, config Config, transport transport.Client, overlay overlay.DB, loop *metainfo.Loop) *Service {
+	return &Service{
+		log:    log,
+		config: config,
+		Loop:   *sync2.NewCycle(config.Interval),
+
+		transport:    transport,
+		overlay:      overlay,
+		metainfoLoop: loop,
+	}
+}
+
+// Run starts the gc loop service
+func (service *Service) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
```

Comment:
If I understand correctly what kind of traces `mon.Task` do I'm wondering how useful is to trace the call to this function because it runs a loop so the trace will be so long and we mostly know when it stops.

I think that it will more useful to trace each iteration of the loop.

Reply:
so do you think I should remove it from here? I am not sure how to trace each iteration of the loop. We do have some stats that we send at the end of each successful iteration

Reply:
I'm not entirely sure about monkit but in other systems you get a span on each function call which is traced which extends through the time from the start to the end; if that's the case, I believe that's more useful to know the time on each iteration rather than the `Run` function.

I think that somebody with more knowledge about monkit should take a look to this.

Reply:
I think it makes sense to extract the anonymous function passed to `service.Loop.Run()` as a separate method and call `mon.Task` there.

Reply:
and should we remove `mon.Task` from here? If the data wouldn't be useful maybe we should take it out

Reply:
yeah don't keep `mon.Task` we aren't finding valuable because they can make the trace files really big and hard to open at times.

Reply:
I didn't extract the anonymous function, but I moved the `defer mon.Task` to inside that function